base/int.jl

```diff
@@ -96,6 +96,44 @@ inv(x::Integer) = float(one(x)) / float(x)
 # skip promotion for system integer types
 (/)(x::BitInteger, y::BitInteger) = float(x) / float(y)
 
+
+"""
+    mul_hi(a::T, b::T) where {T<:Integer}
```

Comment:
```suggestion
    mul_hi(a::T, b::T) where {T<:Base.BitInteger}
```
This doesn't yet support general `Integer`s. 

Alternatively, it would be reasonable to declare it with `T <: Integer` if we defined it for other Base integers (i.e. `Bool`, `BigInt`) or decide that we want mul_hi to throw on BigInt, and specify an extensible API that specifies how this should be extended for custom types (e.g. checked and unbounded).

Reply:
To me it sounds reasonable to restrict to Base.BitInteger because "the higher half of the product" may not make sense for other types.

Reply:
Thanks for the suggestion

Reply:
> specify an extensible API that specifies how this should be extended for custom types

I am not sure.

Making Base.mul_hi extensible can potentially be useful if a package defines Int1024 and implements mul_hi for it.

Then one question is how generic the interface should be.
For example, getting the higher part is useful for polynomial multiplication (perhaps Nemo has this). 
So would it extend mul_hi, and should we have Base.mul_hi(::Any, ::Any).

I do not think a generic interface is needed because I rarely expect my code written specifically for machine integers to also work with wide integers, polynomials, etc.